HugeCTR/include/hps/hash_map_backend.hpp

```diff
@@ -126,21 +156,53 @@ class HashMapBackend final : public VolatileBackend<Key, HashMapBackendParams> {
     std::vector<ValuePage> value_pages;
     std::vector<ValuePtr> value_slots;
 
+    mutable std::shared_mutex read_write_lck;
+
     // Key -> Payload map.
-    phmap::flat_hash_map<Key, Payload> entries;
+    phmap::parallel_flat_hash_map<Key, Payload, phmap::priv::hash_default_hash<Key>,
+                                  phmap::priv::hash_default_eq<Key>,
+                                  phmap::priv::Allocator<phmap::priv::Pair<const Key, Payload>>, 
+                                  4, std::shared_mutex> entries;
```

Comment:
Yes, it will avoid breaking the updates, but partly due to reasons that I wanted to avoid.

Why `parallel_hash_map` does with this configuration is to split the hash table into 4 independent pieces.

The submap is this  thing in their code:
```
struct Inner : public Lockable
```

Upon query each call ends up with a structure like this,
```
typename Lockable::UniqueLock m;
auto res = this->find_or_prepare_insert(k, m);
```
which in turn executes:
```
Inner& inner = sets_[subidx(hashval)];
auto&  set   = inner.set_;
mutexlock    = std::move(typename Lockable::UniqueLock(inner));
```

. So in other words. It will select one of the 4 independent submaps and FULLY lock it. So there is no parallelism possible on that submap. So in the very BEST case when you have 4 keys, each being hashed by `subidx` to different sub-hashmaps, you can accomodate up to 4 threads. Arguably, that is better than now. But you need to be aware this the worst-case still exists. Hence the reason why we didn't use their parallelized hashmap implementation yet.

Reply:
@yingcanw We can take over this part of the change without much issues. In my new streamlined hashmap we will use a similar method to speed things up.